security: don't exceed CA expiration for server/client certs.

mberhault opened the pull request:
It's generally not a good idea to have the server/client certs exceed
the CA lifetime.
This will also make monitoring more accurate, avoiding the need to
monitor certificate chains (although people could still craft their
own).

Commit:
security: don't exceed CA expiration for server/client certs.

It's generally not a good idea to have the server/client certs exceed
the CA lifetime.
This will also make monitoring more accurate, avoiding the need to
monitor certificate chains (although people could still craft their
own).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16055)
<!-- Reviewable:end -->

mberhault commented on pkg/security/x509.go:
```diff
@@ -51,8 +54,9 @@ func newTemplate(commonName string, lifetime time.Duration) (*x509.Certificate,
 
-	notBefore := timeutil.Now().Add(validFrom)
-	notAfter := notBefore.Add(lifetime)
+	now := timeutil.Now()
```
the extra padding (see reason on L40) meant that users weren't getting the lifetime they expected. I switched to using the actual expiration date and padding the "NotBefore" date.

mberhault commented:
@bdarnell: this addresses your comment from the cert rotation RFC (#14254):
`
A CA should not issue certificates that expire after the CA itself expires (do our cert tools enforce this?), so if we instruct users to deploy a new CA cert to clients whenever they deploy new client keypairs, then monitoring the expiration of the client certs should be good enough (or at least as good as we can get on the cockroach side).
`

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-KkkyGlywk93o9dDxW31:bjsulj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
> ```Go
> 	if template.NotAfter.After(caCert.NotAfter) {
> 		log.Infof(context.Background(), "Using expiration date from CA certificate: %s", caCert.NotAfter)
> 		template.NotAfter = caCert.NotAfter
> ```

Silently shortening the life of a cert could lead to surprises. I think it's better to return an error than to change the duration of the cert (although this requires changing the constants so that CAs live at least twice as long as regular certs). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkl1M1Xfu0JvAzOZXTX:b-va8dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Silently shortening the life of a cert could lead to surprises. I think it's better to return an error than to change the duration of the cert (although this requires changing the constants so that CAs live at least twice as long as regular certs). 
</blockquote></details>

Ok, but then we'll want an override flag. If you're asking for 10 years and the CA has 9.5 years left, that still seems reasonable.
Another alternative to different lifetimes would be if the requested lifetime gets shrunk more than X percent (eg: 25% or 50%). Not sure of a good threshold here though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkl2JYo0_9Y3HaPdZDF:b-92hx4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Ok, but then we'll want an override flag. If you're asking for 10 years and the CA has 9.5 years left, that still seems reasonable.
Another alternative to different lifetimes would be if the requested lifetime gets shrunk more than X percent (eg: 25% or 50%). Not sure of a good threshold here though.
</blockquote></details>

Isn't the override flag just `--lifetime=whatever-is-available`? 

We're using 10 years to mean "~forever", so shortening from 10 years might make sense. But other than that, when people ask for a cert with a 1-year lifespan they want to be able to put a reminder on their calendars a year in the future and not have to think about it until then, so it's important to give them at least as much time as they ask for. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkl5GX4JzH3PMnHBkN4:b-ky30x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Isn't the override flag just `--lifetime=whatever-is-available`? 

We're using 10 years to mean "~forever", so shortening from 10 years might make sense. But other than that, when people ask for a cert with a 1-year lifespan they want to be able to put a reminder on their calendars a year in the future and not have to think about it until then, so it's important to give them at least as much time as they ask for. 
</blockquote></details>

The issue with just failing is that it will make the tool pretty much always unusable for automation with default flags since default lifetimes are the same. We could go back to mismatched lifetimes but I feel we've had this argument already.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkl6hQ1Aqp90gMo5hu5:b-7kj3ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

The issue with just failing is that it will make the tool pretty much always unusable for automation with default flags since default lifetimes are the same. We could go back to mismatched lifetimes but I feel we've had this argument already.
</blockquote></details>

Yeah, that's why I suggested changing the constants in my previous comment. We went from 2/5 to 10/10 without thinking about the fact that that implies that we'll always try to generate a cert that outlives the CA. I think it's better to switch to 5/10 or 10/20 than to have a rule that we can generate certs with a shorter lifespan than requested. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkl7rF0yFzz0nbJ-6eZ:bmnin1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, that's why I suggested changing the constants in my previous comment. We went from 2/5 to 10/10 without thinking about the fact that that implies that we'll always try to generate a cert that outlives the CA. I think it's better to switch to 5/10 or 10/20 than to have a rule that we can generate certs with a shorter lifespan than requested. 
</blockquote></details>

ok. I'd be happy with 10/20. I'll fix this up.
@a-robinson: thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkl9RP_ke4o4uodjWLo:bl1v9ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

ok. I'd be happy with 10/20. I'll fix this up.
@a-robinson: thoughts?
</blockquote></details>

SGTM if it'll help with the tooling. I don't think too many people will be confused by having to replace their node certs when they have to replace their CA certs 10 years down the road, but 10/20 is fine with me too

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-KklC4uMc0xFeEkOhS5p:b-umnrp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

SGTM if it'll help with the tooling. I don't think too many people will be confused by having to replace their node certs when they have to replace their CA certs 10 years down the road, but 10/20 is fine with me too
</blockquote></details>

The problem isn't just replacing the node certs when they replace the CA certs, but adding new nodes 6 months from now (assuming they're generating per-node certs instead of creating one and reusing it). You need plenty of room on the CA certs to allow node certs (and user certs, etc) to be issued after the initial setup. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

mberhault commented:
PTAL: switched to failing the certificate creation and printing the time in a format that is usable as a flag.
eg:
```
$ cockroach cert --logtostderr=INFO create-client --ca-key=${HOME}/.cockroach-certs/ca.key foo --overwrite 
Error: failed to generate client certificate and key: error creating client certificate and key: CA lifetime is 87328.698573h, shorter than the requested 87840.000000h
```

I've intentionally kept it as a float as opposed to truncating to hours so that you don't get confused by if the expiration is only minutes off.

mberhault commented:
btw: the 20/10 may not be a great idea. Anyone who has already generated a CA will get this error (since they'll be at 10 years). It may be better to switch to 10/5.

bdarnell commented:
LGTM. Good point about switching to 10/5.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/security/x509.go, line 110 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkluMiKPUcx5MeeiG_s:-KkluMiKPUcx5MeeiG_t:bytcd6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6f1a1ee4798aa1633fe436885036d36066440b/pkg/security/x509.go#L110)):*
> ```Go
> 	niceCALifetime := ca.NotAfter.Sub(now).Hours()
> 	niceCertLifetime := cert.NotAfter.Sub(now).Hours()
> 	return errors.Errorf("CA lifetime is %fh, shorter than the requested %fh",
> ```

Hmm, hours is not a very readable format, but I guess that's the best we can do here. Maybe point to the flag to use: `"Try again with --lifetime=%fh to use the shorter duration"` (round down to something that will work for at least a few minutes)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-:-KklvYdBPHh0CCvIhLTo:b8pfe9d)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Switched to 10/5 years for CA/Others respectively.
Tweaked the error message a bit.
Also fixed the few tests that generate their own certs to have a longer CA lifetime.

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/security/x509.go, line 110 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkluMiKPUcx5MeeiG_s:-Kkm6QV2HewcxzwTz2in:b-952kgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6f1a1ee4798aa1633fe436885036d36066440b/pkg/security/x509.go#L110)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Hmm, hours is not a very readable format, but I guess that's the best we can do here. Maybe point to the flag to use: `"Try again with --lifetime=%fh to use the shorter duration"` (round down to something that will work for at least a few minutes)
</blockquote></details>

Done. here's the new message:
```
$ ./cockroach cert create-client --ca-key=${HOME}/.cockroach-certs/ca.key foo --lifetime=87762h
Error: failed to generate client certificate and key: error creating client certificate and key: CA lifetime is 87761.500375h, shorter than the requested 87762.000000h. Renew CA certificate, or rerun with --lifetime=87761h for a shorter duration.
Failed running "cert"
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-:-Kkm6kxNVG4frInBiSXP:blt97x5)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 1 of 1 files at r3.
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-KkmDG6djv6viB4gl2aA:b-9oakaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The problem isn't just replacing the node certs when they replace the CA certs, but adding new nodes 6 months from now (assuming they're generating per-node certs instead of creating one and reusing it). You need plenty of room on the CA certs to allow node certs (and user certs, etc) to be issued after the initial setup. 
</blockquote></details>

Ah, I misunderstood the conversation earlier. Dropping node cert lifetimes down to 5 years makes me sad, since unsuspecting users will feel pain twice as soon as before this change.

I agree erroring out makes sense if a lifetime was explicitly specified by the user, but can't we just print a warning to users if they don't specify it and we have to reduce the default generated cert lifetime?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-KkmMp_zpBlmJwL5JvmA:bbbbkew) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Ah, I misunderstood the conversation earlier. Dropping node cert lifetimes down to 5 years makes me sad, since unsuspecting users will feel pain twice as soon as before this change.

I agree erroring out makes sense if a lifetime was explicitly specified by the user, but can't we just print a warning to users if they don't specify it and we have to reduce the default generated cert lifetime?
</blockquote></details>

I hope within 5 years we'll have a better cert generation/distribution/rotation solution and we won't have to rely on the hack of very long lifetimes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-KkmnWEf6f7qi8K_r4bv:b-b9h9gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I hope within 5 years we'll have a better cert generation/distribution/rotation solution and we won't have to rely on the hack of very long lifetimes.
</blockquote></details>

For node certs, yeah, but client certs are partially out of our control

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/security/x509.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055#-KkkyGlywk93o9dDxW30:-Kkq7Xva4VeT0c6EtMwg:b-1prhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7769be0dbb14b38ff977e26961af76273b2ae23/pkg/security/x509.go#L123)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

For node certs, yeah, but client certs are partially out of our control
</blockquote></details>

True, but I think we can add some monitoring for client certificates as well. It won't be as good since we need to actually get requests from said clients and we most likely be won't be able to integrate client cert deployment with whatever we're doing, but I think it should be sufficient.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16055)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Have we reached some sort of agreement, or at we at an impasse?

a-robinson commented:
Eh, you know my opinion (https://github.com/cockroachdb/cockroach/pull/16055#issuecomment-303235565) and you know @bdarnell's opinion (https://github.com/cockroachdb/cockroach/pull/16055#issuecomment-303161782). You should just pick something and move on. Neither choice will destroy our product, just generate some user questions/complaints 4-10 years down the road.

mberhault commented:
ok, let's stick with 10/5. Between the existing monitoring and debugging, further monitoring (#16094) and some yet-TBD cert deployment improvements, I feel it's safe enough.
